src/main/java/randoop/output/FailingTestFilter.java

```diff
@@ -0,0 +1,235 @@
+ * state effects due to files in the working directory.
+ */
+public class FailingTestFilter implements CodeWriter {
+
+  /** A pattern matching the JUnit4 message indicating the total count of failures */
+  private static final Pattern FAILURE_MESSAGE_PATTERN =
+      Pattern.compile("There\\s+(?:was|were)\\s+(\\d+)\\s+failure(?:s|):");
+
+  /** Regex for Java identifiers */
+  private static final String ID_STRING =
+      "\\p{javaJavaIdentifierStart}\\p{javaJavaIdentifierPart}*";
+
+  /** The {@link randoop.execution.TestEnvironment} for running the test classes. */
+  private final TestEnvironment testEnvironment;
+
+  /** The underlying {@link randoop.output.JavaFileWriter} for writing a test class. */
+  private final JavaFileWriter javaFileWriter;
+
+  /**
+   * Create a {@link FailingTestFilter} run in the environment and using the given {@link
+   * JavaFileWriter} to output test classes.
+   *
+   * @param testEnvironment the {@link TestEnvironment} for executing tests during filtering
+   * @param javaFileWriter the {@link JavaFileWriter} to write java files for the classes
+   */
+  public FailingTestFilter(TestEnvironment testEnvironment, JavaFileWriter javaFileWriter) {
+    this.testEnvironment = testEnvironment;
+    this.javaFileWriter = javaFileWriter;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Inserts comments in place of assertions that fail when the test is run.
+   *
+   * <p>May be sensitive to changes in JUnit runner output.
+   */
+  @Override
+  public File writeClass(String packageName, String classname, String classString) {
+
+    List<File> sourceList = new ArrayList<>();
+    File testFile = javaFileWriter.writeClass(packageName, classname, classString);
+    sourceList.add(testFile);
+
+    String qualifiedClassname = (packageName.isEmpty() ? "" : packageName + ".") + classname;
+
+    int pass = 0;
+    boolean passing = false;
+
+    while (!passing) {
+
+      Path workingDirectory = createWorkingDirectory(classname, pass);
+
+      boolean success = compileTestClass(sourceList, workingDirectory);
+      assert success;
+
+      RunStatus status = null;
+      try {
+        status = testEnvironment.runTest(qualifiedClassname, workingDirectory.toFile());
+      } catch (ProcessException e) {
+        throw new BugInRandoopException("Error filtering regression tests: " + e.getMessage());
+      }
+      if (status.exitStatus == 0) {
+        passing = true;
+        continue;
+      }
+
+      /*
+       * Read the standard output from running the test class using JUnit4.
+       */
+      assert status.errorOutputLines.isEmpty()
+          : "JUnit writes to standard out, error output unexpected";
+
+      /*
+       * First, find the message that indicates the number of failures in the run.
+       */
+      int totalFailures = -1;
+      Iterator<String> lineIterator = status.standardOutputLines.iterator();
+      while (lineIterator.hasNext() && totalFailures < 0) {
```

Comment:
This way of writing a while loop is confusing.  I would put a `break;` statement after the assignment to `totalFailures`, which clarifies the exit condition at the relevant place.  Then you don't need the second clause at all.